File: packages/csharp/SourceRpc.Continuity/Manifest.cs
namespace SourceRpc.Continuity;

/// <summary>
/// What a compiled artifact says about itself, and what has to be checked against it.
///
/// A port of `Manifest.ts`. Until cross-language handoff, a replacement was checked by the compiler
/// that built both sides; here nothing is checked by anything unless it is written down. So the
/// artifact carries a manifest, and every function in this file takes it as a *claim*.
///
/// **The manifest describes the revision. It does not grant authority.** A manifest is emitted by
/// the artifact, and an artifact that could authorise itself by asserting its own capabilities would
/// make the whole approval path decorative.
/// </summary>
public enum RpcArtifactType
{
    JavaScript,
    DotNet,
    Wasm,
    Native,
    SourceIr
}

/// <summary>
/// What the artifact claims about its own suitability for online change.
///
/// Every field is a claim about the *code*, which is why none can be verified by a runtime and all
/// have to be declared. <see cref="SerialisedHandlers"/> is the one the barrier rests on: a component
/// whose state is changed by anything other than a handler the runtime dispatched is not quiescent
/// when the queue is empty, and no barrier can detect it.
/// </summary>
public sealed record RpcOnlineChangeProfile
{
    public required bool Supported { get; init; }
    public required bool SerialisedHandlers { get; init; }
    public required bool RuntimeManagedObligations { get; init; }
    public required int QuiescenceDeadlineMs { get; init; }
}

public sealed record RpcContractReference
{
    public required string Id { get; init; }
    public required int Version { get; init; }
    public required string SchemaHash { get; init; }
}

public sealed record RpcStateReference
{
    public required string SchemaId { get; init; }
    public required int Version { get; init; }
    public required string SchemaHash { get; init; }
}

public sealed record RpcRevisionManifest
{
    public required int ManifestVersion { get; init; }
    public required string ComponentType { get; init; }
    public required string RevisionId { get; init; }
    public required RpcArtifactType ArtifactType { get; init; }

    /// <summary>The digest of the built artifact, taken by whatever built it. Compared, never recomputed here.</summary>
    public required string ArtifactHash { get; init; }

    public required RpcContractReference Contract { get; init; }
    public required RpcStateReference State { get; init; }
    public IReadOnlyList<string> RequiredCapabilities { get; init; } = [];
    public required RpcOnlineChangeProfile OnlineChange { get; init; }
    public required string ManifestHash { get; init; }
}

/// <summary>What a deployment has decided a component may be, independently of what any artifact says.</summary>
public sealed record RpcIdentityPolicy
{
    public required string ComponentId { get; init; }
    public required string ComponentType { get; init; }

    /// <summary>The artifact digests approved for this identity. An empty list approves nothing, not everything.</summary>
    public IReadOnlyList<string> ApprovedArtifacts { get; init; } = [];

    /// <summary>The most this identity may ever be granted, whatever a revision asks for.</summary>
    public IReadOnlyList<string> CapabilityEnvelope { get; init; } = [];

    /// <summary>Whether this identity may be changed while running at all. Some plant is not eligible, by decision.</summary>
    public required bool OnlineChangePermitted { get; init; }
}

/// <summary>Whether this revision agrees with a snapshot, and whether the migration chain has to run.</summary>
public sealed record RpcReconciliation
{
    public required bool Agreed { get; init; }
    public bool MigrationNeeded { get; init; }
    public string? Why { get; init; }
}

public static class RpcManifests
{
    /// <summary>The manifest version this implementation writes and the highest it reads.</summary>
    public const int Version = 1;

    /// <summary>Recompute the manifest hash. The reason it does not match, or nothing.</summary>
    public static string? Verify(RpcRevisionManifest manifest)
    {
        var expected = RpcCanonicalDigest.Digest(HashedForm(manifest));
        return expected == manifest.ManifestHash ? null : $"manifest for {manifest.RevisionId} hashes to {expected}, and carries {manifest.ManifestHash}";
    }

    /// <summary>
    /// Whether a revision can take over a particular snapshot.
    ///
    /// This is the check that makes cross-language handoff mean something. The two artifacts share
    /// no compiler, no type system and no runtime; what they share is a component type, a contract
    /// hash and a state schema hash, and if those agree then the successor holds the same description
    /// of the same values that the incumbent did. If they do not, nothing else about the two being
    /// interface-compatible matters.
    ///
    /// The state *version* may legitimately differ - that is what migration is for - so a mismatch
    /// there is reported separately from a mismatch of identity, which is never migratable.
    /// </summary>
    public static RpcReconciliation Reconcile(RpcRevisionManifest manifest, RpcSnapshotEnvelope snapshot)
    {
        if (manifest.ComponentType != snapshot.ComponentType)
            return Refuse($"{manifest.RevisionId} implements {manifest.ComponentType} and this snapshot is of {snapshot.ComponentType}: two component types are not two versions of one");
        if (manifest.State.SchemaId != snapshot.StateSchemaId)
            return Refuse(
                $"{manifest.RevisionId} holds {manifest.State.SchemaId} and this snapshot carries {snapshot.StateSchemaId}: a schema id is stable for the life of a component type, so two of them are two different states");
        if (manifest.State.Version == snapshot.StateVersion && manifest.State.SchemaHash != snapshot.StateSchemaHash)
            return Refuse(
                $"{manifest.RevisionId} and this snapshot both claim {manifest.State.SchemaId} v{manifest.State.Version} and describe it differently ({manifest.State.SchemaHash} against {snapshot.StateSchemaHash}): a published version cannot be redefined, and one of these two was");
        if (manifest.State.Version < snapshot.StateVersion)
            return Refuse($"{manifest.RevisionId} holds {manifest.State.SchemaId} v{manifest.State.Version} and this snapshot is at v{snapshot.StateVersion}: migration is forward only, and this would be a rollback");
        return new RpcReconciliation { Agreed = true, MigrationNeeded = manifest.State.Version != snapshot.StateVersion };
    }

    /// <summary>
    /// Whether this artifact is allowed to be this component.
    ///
    /// The reason it is not, and there are several because they are several different conversations:
    /// the wrong type is a mistake, an unapproved artifact needs a deployment approval, a capability
    /// outside the envelope needs the envelope widened by whoever owns the identity, and an identity
    /// that is not eligible for online change needs a controlled restart instead. Collapsing them
    /// into `false` would leave every one of those as the same shrug.
    /// </summary>
    public static string? Authorised(RpcRevisionManifest manifest, RpcIdentityPolicy policy)
    {
        if (manifest.ComponentType != policy.ComponentType) return $"{policy.ComponentId} is a {policy.ComponentType} and {manifest.RevisionId} implements {manifest.ComponentType}";
        if (!policy.ApprovedArtifacts.Contains(manifest.ArtifactHash))
            return $"{manifest.RevisionId} ({manifest.ArtifactHash}) is not among the artifacts approved for {policy.ComponentId}: a manifest describes a revision, it does not approve one";
        var beyond = manifest.RequiredCapabilities.Where(capability => !policy.CapabilityEnvelope.Contains(capability)).ToList();
        if (beyond.Count > 0)
            return $"{manifest.RevisionId} requires {string.Join(", ", beyond)}, which {policy.ComponentId} is not permitted to grant - an interface-compatible replacement does not inherit an authority the identity never had";
        if (!policy.OnlineChangePermitted) return $"{policy.ComponentId} is not eligible for online change, so {manifest.RevisionId} is deployed by a controlled restart rather than a handoff";
        if (!manifest.OnlineChange.Supported) return $"{manifest.RevisionId} does not support online change and says so in its own manifest";
        if (!manifest.OnlineChange.SerialisedHandlers)
            return $"{manifest.RevisionId} does not serialise its handlers, so no barrier can establish that it is quiescent - the queue being empty would say nothing about what is running";
        return null;
    }

    private static RpcReconciliation Refuse(string why) => new() { Agreed = false, Why = why };

    internal static IReadOnlyDictionary<string, object?> HashedForm(RpcRevisionManifest manifest) =>
        new Dictionary<string, object?>
        {
            ["manifestVersion"] = (double)manifest.ManifestVersion,
            ["componentType"] = manifest.ComponentType,
            ["revisionId"] = manifest.RevisionId,
            ["artifactType"] = ArtifactName(manifest.ArtifactType),
            ["artifactHash"] = manifest.ArtifactHash,
            ["contract"] = new Dictionary<string, object?> { ["id"] = manifest.Contract.Id, ["version"] = (double)manifest.Contract.Version, ["schemaHash"] = manifest.Contract.SchemaHash },
            ["state"] = new Dictionary<string, object?>
            {
                ["schemaId"] = manifest.State.SchemaId,
                ["version"] = (double)manifest.State.Version,
                ["schemaHash"] = manifest.State.SchemaHash
            },
            // Sorted, because a manifest listing the same capabilities in a different order is the
            // same manifest, and an artifact rebuilt on a machine that walked its imports
            // differently would otherwise hash to something else and read as a different revision.
            ["requiredCapabilities"] = manifest.RequiredCapabilities.OrderBy(one => one, StringComparer.Ordinal).ToList(),
            ["onlineChange"] = new Dictionary<string, object?>
            {
                ["supported"] = manifest.OnlineChange.Supported,
                ["serialisedHandlers"] = manifest.OnlineChange.SerialisedHandlers,
                ["runtimeManagedObligations"] = manifest.OnlineChange.RuntimeManagedObligations,
                ["quiescenceDeadlineMs"] = (double)manifest.OnlineChange.QuiescenceDeadlineMs
            }
        };

    /// <summary>The wire spelling, as a switch rather than a member name, for <see cref="RpcNames"/>'s reason.</summary>
    public static string ArtifactName(RpcArtifactType type) =>
        type switch
        {
            RpcArtifactType.JavaScript => "javascript",
            RpcArtifactType.DotNet => "dotnet",
            RpcArtifactType.Wasm => "wasm",
            RpcArtifactType.Native => "native",
            RpcArtifactType.SourceIr => "source-ir",
            _ => throw new ArgumentOutOfRangeException(nameof(type))
        };

    /// <summary>Read a manifest another language wrote. Refuses rather than defaulting, as everything here does.</summary>
    public static RpcRevisionManifest Read(string json)
    {
        using var document = System.Text.Json.JsonDocument.Parse(json);
        var root = document.RootElement;
        var version = root.GetProperty("manifestVersion").GetInt32();
        if (version > Version) throw new RpcPortableSnapshot.RefusedException($"this manifest is version {version} and this implementation reads up to {Version}", "manifestVersion");
        var contract = root.GetProperty("contract");
        var state = root.GetProperty("state");
        var online = root.GetProperty("onlineChange");
        return new RpcRevisionManifest
        {
            ManifestVersion = version,
            ComponentType = root.GetProperty("componentType").GetString()!,
            RevisionId = root.GetProperty("revisionId").GetString()!,
            ArtifactType = ArtifactType(root.GetProperty("artifactType").GetString()!),
            ArtifactHash = root.GetProperty("artifactHash").GetString()!,
            Contract = new RpcContractReference
            {
                Id = contract.GetProperty("id").GetString()!,
                Version = contract.GetProperty("version").GetInt32(),
                SchemaHash = contract.GetProperty("schemaHash").GetString()!
            },
            State = new RpcStateReference
            {
                SchemaId = state.GetProperty("schemaId").GetString()!,
                Version = state.GetProperty("version").GetInt32(),
                SchemaHash = state.GetProperty("schemaHash").GetString()!
            },
            RequiredCapabilities = root.TryGetProperty("requiredCapabilities", out var capabilities) ? capabilities.EnumerateArray().Select(one => one.GetString()!).ToList() : [],
            OnlineChange = new RpcOnlineChangeProfile
            {
                Supported = online.GetProperty("supported").GetBoolean(),
                SerialisedHandlers = online.GetProperty("serialisedHandlers").GetBoolean(),
                RuntimeManagedObligations = online.GetProperty("runtimeManagedObligations").GetBoolean(),
                QuiescenceDeadlineMs = online.GetProperty("quiescenceDeadlineMs").GetInt32()
            },
            ManifestHash = root.GetProperty("manifestHash").GetString()!
        };
    }

    private static RpcArtifactType ArtifactType(string value) =>
        value switch
        {
            "javascript" => RpcArtifactType.JavaScript,
            "dotnet" => RpcArtifactType.DotNet,
            "wasm" => RpcArtifactType.Wasm,
            "native" => RpcArtifactType.Native,
            "source-ir" => RpcArtifactType.SourceIr,
            _ => throw new RpcPortableSnapshot.RefusedException($"artifactType is \"{value}\", which this implementation does not know", "artifactType")
        };
}

File: packages/continuity/src/Manifest.ts
import { canonicalText, digestText } from '@source-repo/rpc'
import { RpcSnapshotRefused, type RpcSnapshotEnvelope } from './Envelope.js'

/**
 * What a compiled artifact says about itself.
 *
 * Until now a handoff was between two things this process already knew - two classes in one
 * language, checked by the compiler that built both. Across languages nothing is checked by anything
 * unless it is written down, so the artifact carries a manifest: what component type it implements,
 * which contract at which version, which state schema at which version, what it needs to be allowed
 * to do, and whether it can be changed online at all.
 *
 * **The manifest describes the revision. It does not grant authority.** That sentence is the one
 * rule this file exists to enforce, and it is easy to lose: a manifest is emitted by the artifact,
 * and an artifact that could authorise itself by asserting its own capabilities would make the whole
 * approval path decorative. Every check here takes the manifest as a *claim* and something else - an
 * approval record, a snapshot, an identity policy - as the thing it is checked against.
 */

/**
 * What kind of thing this is.
 *
 * Present so the manifest can be read without knowing what produced it, which is the situation the
 * whole phase is about: a coordinator choosing between a JavaScript and a .NET replacement should
 * not have to infer which it is holding from a file extension.
 */
export type RpcArtifactType = 'javascript' | 'dotnet' | 'wasm' | 'native' | 'source-ir'

/**
 * What the artifact claims about its own suitability for online change.
 *
 * Every field is a claim about the *code*, which is why none of them can be verified by the runtime
 * and all of them have to be declared. `serialisedHandlers` is the one the barrier depends on: a
 * component whose state is changed by anything other than a handler the runtime dispatched is not
 * quiescent when the queue is empty, and no barrier can detect it. Phase 2 recorded that limit in a
 * test; this is where a revision states which side of it it is on.
 */
export interface RpcOnlineChangeProfile {
    readonly supported: boolean
    readonly serialisedHandlers: boolean
    readonly runtimeManagedObligations: boolean
    readonly quiescenceDeadlineMs: number
}

export interface RpcRevisionManifest {
    readonly manifestVersion: number
    readonly componentType: string
    readonly revisionId: string
    readonly artifactType: RpcArtifactType
    /** The digest of the built artifact, taken by whatever built it. Compared, never recomputed here. */
    readonly artifactHash: string
    readonly contract: { readonly id: string; readonly version: number; readonly schemaHash: string }
    readonly state: { readonly schemaId: string; readonly version: number; readonly schemaHash: string }
    readonly requiredCapabilities: readonly string[]
    readonly onlineChange: RpcOnlineChangeProfile
    /** The digest of everything above. Set by `sealManifest`. */
    readonly manifestHash: string
}

/** The version this implementation writes and the highest it reads. */
export const RPC_MANIFEST_VERSION = 1

const hashedForm = (draft: Omit<RpcRevisionManifest, 'manifestHash'>) => ({
    manifestVersion: draft.manifestVersion,
    componentType: draft.componentType,
    revisionId: draft.revisionId,
    artifactType: draft.artifactType,
    artifactHash: draft.artifactHash,
    contract: draft.contract,
    state: draft.state,
    // Sorted, because a manifest listing the same capabilities in a different order is the same
    // manifest, and an artifact rebuilt on a machine that walked its imports differently would
    // otherwise hash to something else and read as a different revision.
    requiredCapabilities: [...draft.requiredCapabilities].sort(),
    onlineChange: draft.onlineChange
})

const required = ['componentType', 'revisionId', 'artifactType', 'artifactHash'] as const

/** Name the manifest by its content, refusing one that cannot say what it is. */
export const sealManifest = async (draft: Omit<RpcRevisionManifest, 'manifestHash' | 'manifestVersion'> & { manifestVersion?: number }): Promise<RpcRevisionManifest> => {
    for (const field of required) if (!draft[field]) throw new RpcSnapshotRefused(`a revision manifest states its ${field}`, field)
    if (!draft.contract?.schemaHash) throw new RpcSnapshotRefused('a revision manifest states the hash of the contract it implements', 'contract.schemaHash')
    if (!draft.state?.schemaHash) throw new RpcSnapshotRefused('a revision manifest states the hash of the state schema it holds', 'state.schemaHash')
    if (draft.onlineChange.supported && draft.onlineChange.quiescenceDeadlineMs <= 0)
        throw new RpcSnapshotRefused('a revision that supports online change states a quiescence deadline, because a barrier with no deadline is a component that stops answering for good', 'onlineChange.quiescenceDeadlineMs')

    const sealed = { ...draft, manifestVersion: draft.manifestVersion ?? RPC_MANIFEST_VERSION, requiredCapabilities: [...draft.requiredCapabilities].sort() }
    return Object.freeze({ ...sealed, manifestHash: await digestText(canonicalText(hashedForm(sealed))) })
}

/** Recompute the hash. The reason it does not match, or nothing. */
export const verifyManifest = async (manifest: RpcRevisionManifest): Promise<string | undefined> => {
    const expected = await digestText(canonicalText(hashedForm(manifest)))
    return expected === manifest.manifestHash ? undefined : `manifest for ${manifest.revisionId} hashes to ${expected}, and carries ${manifest.manifestHash}`
}

/**
 * Whether a revision can take over a particular snapshot.
 *
 * This is the check that makes cross-language handoff mean something. The two artifacts share no
 * compiler, no type system and no runtime; what they share is a component type, a contract hash and
 * a state schema hash, and if those agree then the successor is holding the same description of the
 * same values that the incumbent was. If they do not, nothing else about the two being
 * interface-compatible matters.
 *
 * The state *version* may legitimately differ - that is what migration is for - so a mismatch there
 * is reported separately from a mismatch of identity, which is never migratable.
 */
export const reconcile = (manifest: RpcRevisionManifest, snapshot: RpcSnapshotEnvelope): { readonly agreed: true; readonly migrationNeeded: boolean } | { readonly agreed: false; readonly why: string } => {
    if (manifest.componentType !== snapshot.componentType)
        return { agreed: false, why: `${manifest.revisionId} implements ${manifest.componentType} and this snapshot is of ${snapshot.componentType}: two component types are not two versions of one` }
    if (manifest.state.schemaId !== snapshot.stateSchemaId)
        return { agreed: false, why: `${manifest.revisionId} holds ${manifest.state.schemaId} and this snapshot carries ${snapshot.stateSchemaId}: a schema id is stable for the life of a component type, so two of them are two different states` }
    if (manifest.state.version === snapshot.stateVersion && manifest.state.schemaHash !== snapshot.stateSchemaHash)
        return {
            agreed: false,
            why: `${manifest.revisionId} and this snapshot both claim ${manifest.state.schemaId} v${manifest.state.version} and describe it differently (${manifest.state.schemaHash} against ${snapshot.stateSchemaHash}): a published version cannot be redefined, and one of these two was`
        }
    if (manifest.state.version < snapshot.stateVersion)
        return { agreed: false, why: `${manifest.revisionId} holds ${manifest.state.schemaId} v${manifest.state.version} and this snapshot is at v${snapshot.stateVersion}: migration is forward only, and this would be a rollback` }
    return { agreed: true, migrationNeeded: manifest.state.version !== snapshot.stateVersion }
}

/**
 * What a deployment has decided a component may be, independently of what any artifact says.
 *
 * The four concerns the design keeps apart meet here: this is identity policy and artifact
 * authorisation, and the manifest is the claim being measured against them.
 */
export interface RpcIdentityPolicy {
    readonly componentId: string
    readonly componentType: string
    /** The artifact digests approved for this identity. An empty list approves nothing, not everything. */
    readonly approvedArtifacts: readonly string[]
    /** The most this identity may ever be granted, whatever a revision asks for. */
    readonly capabilityEnvelope: readonly string[]
    /** Whether this identity may be changed while running at all. Some plant is not eligible, by decision. */
    readonly onlineChangePermitted: boolean
}

/**
 * Whether this artifact is allowed to be this component.
 *
 * Returns the reason it is not, and there are four of them because they are four different
 * conversations: the wrong type is a mistake, an unapproved artifact needs a deployment approval, a
 * capability outside the envelope needs the envelope widened by whoever owns the identity, and an
 * identity that is not eligible for online change needs a controlled restart instead. Collapsing
 * them into `false` would leave every one of those as the same shrug.
 */
export const authorised = (manifest: RpcRevisionManifest, policy: RpcIdentityPolicy): string | undefined => {
    if (manifest.componentType !== policy.componentType)
        return `${policy.componentId} is a ${policy.componentType} and ${manifest.revisionId} implements ${manifest.componentType}`
    if (!policy.approvedArtifacts.includes(manifest.artifactHash))
        return `${manifest.revisionId} (${manifest.artifactHash}) is not among the artifacts approved for ${policy.componentId}: a manifest describes a revision, it does not approve one`
    const beyond = manifest.requiredCapabilities.filter((capability) => !policy.capabilityEnvelope.includes(capability))
    if (beyond.length)
        return `${manifest.revisionId} requires ${beyond.join(', ')}, which ${policy.componentId} is not permitted to grant - an interface-compatible replacement does not inherit an authority the identity never had`
    if (!policy.onlineChangePermitted) return `${policy.componentId} is not eligible for online change, so ${manifest.revisionId} is deployed by a controlled restart rather than a handoff`
    if (!manifest.onlineChange.supported) return `${manifest.revisionId} does not support online change and says so in its own manifest`
    if (!manifest.onlineChange.serialisedHandlers)
        return `${manifest.revisionId} does not serialise its handlers, so no barrier can establish that it is quiescent - the queue being empty would say nothing about what is running`
    return undefined
}
